File: dashboard/src/core/plugins/inspect_traffic.ts
/*
    proxy.py
    ~~~~~~~~
    ⚡⚡⚡ Fast, Lightweight, Pluggable, TLS interception capable proxy server focused on
    Network monitoring, controls & Application development, testing, debugging.

    :copyright: (c) 2013-present by Abhinav Singh and contributors.
    :license: BSD, see LICENSE for more details.
*/
import { DashboardPlugin } from '../plugin'

export class InspectTrafficPlugin extends DashboardPlugin {
  public name: string = 'inspect_traffic'
  public title: string = 'Inspect Traffic'

  public initializeTab () : JQuery<HTMLElement> {
    return this.makeTab(this.title, 'fa-binoculars')
  }

  public initializeHeader (): JQuery<HTMLElement> {
    return this.makeHeader(this.title)
  }

  public initializeBody (): JQuery<HTMLElement> {
    return $('')
  }

  public activated (): void {
    this.websocketApi.sendMessage(
      { method: 'enable_inspection' },
      this.handleEvents.bind(this))
    this.ensureIFrame()
  }

  public deactivated (): void {
    this.websocketApi.sendMessage(
      { method: 'disable_inspection' })
  }

  public handleEvents (message: Record<string, any>): void {
    console.log(message)
  }

  private ensureIFrame (): void {
    if ($('#' + this.getDevtoolsIFrameID()).length === 0) {
      $('#' + this.name)
        .children('.app-body')
        .append(
          this.getDevtoolsIFrame()
        )
    }
  }

  private getDevtoolsIFrameID (): string {
    return this.name.concat('_inspector')
  }

  private getDevtoolsIFrame (): JQuery<HTMLElement> {
    return $('<iframe></iframe>')
      .attr('id', this.getDevtoolsIFrameID())
      .attr('height', '100%')
      .attr('width', '100%')
      .attr('padding', '0')
      .attr('margin', '0')
      .attr('frameBorder', '0')
      .attr('scrolling', 'no')
      .attr('src', 'devtools/inspect_traffic.html')
  }
}

File: proxy/http/inspector/inspect_traffic.py
# -*- coding: utf-8 -*-
"""
    proxy.py
    ~~~~~~~~
    ⚡⚡⚡ Fast, Lightweight, Pluggable, TLS interception capable proxy server focused on
    Network monitoring, controls & Application development, testing, debugging.

    :copyright: (c) 2013-present by Abhinav Singh and contributors.
    :license: BSD, see LICENSE for more details.
"""
import json
from typing import TYPE_CHECKING, Any, Dict, List

from proxy.core.event import EventSubscriber
from proxy.common.utils import bytes_
from proxy.http.websocket import WebsocketFrame, WebSocketTransportBasePlugin


if TYPE_CHECKING:   # pragma: no cover
    from ..connection import HttpClientConnection


class InspectTrafficPlugin(WebSocketTransportBasePlugin):
    """Websocket API for inspect_traffic.ts frontend plugin."""

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
        self.subscriber = EventSubscriber(
            self.event_queue,
            callback=lambda event: InspectTrafficPlugin.callback(
                self.client, event,
            ),
        )

    def methods(self) -> List[str]:
        return [
            'enable_inspection',
            'disable_inspection',
        ]

    def handle_message(self, message: Dict[str, Any]) -> None:
        if message['method'] == 'enable_inspection':
            # inspection can only be enabled if --enable-events is used
            if not self.flags.enable_events:
                self.reply({
                    'id': message['id'],
                    'response': 'not enabled',
                })
            else:
                self.subscriber.setup()
                self.reply(
                    {'id': message['id'], 'response': 'inspection_enabled'},
                )
        elif message['method'] == 'disable_inspection':
            self.subscriber.shutdown()
            self.reply({
                'id': message['id'],
                'response': 'inspection_disabled',
            })
        else:
            raise NotImplementedError()

    @staticmethod
    def callback(client: 'HttpClientConnection', event: Dict[str, Any]) -> None:
        event['push'] = 'inspect_traffic'
        client.queue(
            memoryview(
                WebsocketFrame.text(
                    bytes_(
                        json.dumps(event),
                    ),
                ),
            ),
        )
